activerecord/lib/active_record/integration.rb

```diff
@@ -96,8 +96,13 @@ def cache_key(*timestamp_names)
     # Note, this method will return nil if ActiveRecord::Base.cache_versioning is set to
     # +false+ (which it is by default until Rails 6.0).
     def cache_version
-      if cache_versioning && timestamp = try(:updated_at)
-        timestamp.utc.to_s(:usec)
+      return unless cache_versioning
+
+      timestamp = try(:updated_at_before_type_cast)
+      if timestamp && timestamp.is_a?(String) && cache_timestamp_format == :db_usec
```

Comment:
I am not sure that checking if the timestamp is a String is sufficient - if a user set it to a String that wan't in the desired format then the cache_key wouldn't match before and after save. 

In addition database adapters that perform the type casting would no longer get any benefit from this as it will be a `Time` that will get cast to `ActiveSupport::TimeWithZone` and back as part of the the formatting of the time.

I think that it is worth checking if the value is from the database as part of this optimisation.